docs/06-concepts/01-working-with-endpoints.md

```diff
@@ -90,3 +89,112 @@ class ExampleEndpoint extends Endpoint {
+    throw UnimplementedError();
+  }
+}
+```
+
+In the above example `ExcitedGreeter` inherits from `GreeterBase`. Since the base class is marked as `abstract` no client was generated for it, but now all of its visible endpoint methods would be exposed through the sub-class `ExcitedGreeter`.
+
+The sub-class applies the following modifications though:
+
+- `hello` is overriden and augments the super class' implementation with a trailing `!!!`
+- `goodbye` is now exposed through the sub-class, as it was not individually hidden and the sub-class does not further augment it
+- `wave`, which is explictly ignored in the base-class is still not exposed, as the sub-class did not opt into re-exposing it
+- `wavePerson` was explicitly overriden to be ignored. Don't worry about the `throw` in the method implementation, which is only needed to satisfy the compiler. In practice this method can not be called from the client anymore.
+
+### API versioning for breaking changes
+
+Endpoint sub-class can be useful when having to do a breaking change on an endpoint, but you want to keep sharing most of it's implementation with the old endpoint.
+
+Imagine you had a "team" management endpoint where before a user could join if they had an e-mail address ending in the expected domain, but now it should be opened up for anyone to join if they can provide an "invite code".
+
+```dart
+@Deprecated('Use TeamV2Endpoint instead')
+class TeamEndpoint extends Endpoint {
+  Future<TeamInfo> join(Session session) async { … }
+  
+  // many more methods, like `leave` etc.
+}
+
+class TeamV2Endpoint extends TeamEndpoint {
+  @override
+  @ignoreEndpoint
+  Future<TeamInfo> join(Session session) async {
+    throw UnimplementedError();
+  }
+
+  Future<TeamInfo> joinWithCode(Session session, String invitationCode) async {
+    …
+  }
+}
+```
+
+In the above example we created a new `TeamV2` endpoint which hides the `join` method and instead exposes a `joinWithCode` method, plus all the other inherited (and untouched) methods from the parent class.
+
+While we may have liked to re-use the `join` method name, Dart inheritance rules do not allow doing so.
+
+Then in your client you could move all usages from `client.team` to `client.teamV2`, and could eventually remove the old endpoint. Either by marking it with `@ignoreEndpoint` on the class or deleting it and moving the untouched method implementations you want to keep to the V2 endpoint class.
+
+An alternative pattern to consider would be to move all the business logic for an endpoint into a helper class, and then call into that from the endpoint. In case you want to create a V2 version later, you might be able to reuse most of the underlying business logic through that helper class, and don't have to sub-class the old endpoint. This has the added benefit of the endpoint class clearly listing all exposed methods, and you don't have to wonder what you inherit from the base class.
```

Comment:
Personally I think as a general "API evolution" approach the externalized business logic and 2 concrete implementations seem better for the common case.

I think extending endpoints is really more suitable in the module -> app case we discussed, where you do it to minimally modify / extend behavior and not do what otherwise would be breaking changes.